jina/executors/decorators.py

```diff
@@ -247,6 +247,10 @@ def arg_wrapper(*args, **kwargs):
             if b_size is None:
                 return func(*args, **kwargs)
 
+            # catch empty shards
+            if data is None:
```

Comment:
I think is not the place to add the fix, I am not sure the return expected is an `array`. I think it needs to be added at `indexer`

Reply:
Well the crash happens at the batch_iterator level

```
    for b in batch_iterator(data[:total_size], b_size, split_over_axis, yield_slice=yield_slice):
TypeError: 'NoneType' object is not subscriptable
```

And I assume it will be several indexers that will call this.

And when looking at the indexer query https://github.com/jina-ai/jina/blob/efdb11342536426b78d1c5d448c64a8f3634027a/jina/executors/indexers/vector.py#L263

```
        if self.metric not in {'cosine', 'euclidean'} or self.backend == 'scipy':
            dist = self._cdist(keys, self.query_handler)
        elif self.metric == 'euclidean':
            _keys = _ext_A(keys)
            dist = self._euclidean(_keys, self.query_handler)
        elif self.metric == 'cosine':
            _keys = _ext_A(_norm(keys))
            dist = self._cosine(_keys, self.query_handler)
        else:
            raise NotImplementedError(f'{self.metric} is not implemented')
```

I see that I would need to add it for each `if` substatement, right?

Also: still a draft. Need to add the tests.

As for return type: if you look at the function that would return after the `for` we can look at this function https://github.com/jina-ai/jina/blob/f0809e32db805401b2768aef7e8eec9c4ebae83d/jina/executors/decorators.py#L179, which seems to return an array

Reply:
yes, maybe some change also needs to be there for `batch_iterator` but it will also affect the `indexer` executor.

Reply:
but it is not necessaritly an `np.array`, it may also a simple `array` I think

Reply:
I think I was wrong it may be actually a bug at batching level. I am howevrr not sure we can assume it returns an np.ndarray. batching decorator has too many things, if you find a pattern to abstract in another function it can also be good

Reply:
I think you're looking at an older version. After discussing it with @maximilianwerk we said the batching should just raise the error. And it's up to the Indexers / modules to know what sort of data and format to default to

Reply:
I think it is not the way to go, because this is not the data passed but the data returned. I think what batching is responsible is to handle data that comes and goes, if it does not receive any data it should just return None.

raising an error is a big statement in my opinion, it should be a natural thing for batching to handle empty data.

Reply:
I disagree. Batching has no clue about the type of the returned value and cannot guess it. If the type is `Option[Any]`, `None` is fine. Anyhow, if the type is not `Option`, `None` does not make any sense, so either we call the underlying function with an empty `np.ndarray` and let it be handled there or we raise an `Exception`.